pkg/rpcmiddleware/rpcretry/client_interceptors.go

```diff
@@ -0,0 +1,183 @@
+	}
+}
+
+// StreamClientInterceptor returns a new streaming client interceptor that retries the execution of external gRPC
+// calls, the retry attempt will only occur if any of the validators define the error as a trigger.
+func StreamClientInterceptor(opts ...Option) grpc.StreamClientInterceptor {
+	callOpts := evaluateClientOpt(opts...)
+	return func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		if callOpts.max == 0 {
+			return streamer(ctx, desc, cc, method, opts...)
+		}
+		if desc.ClientStreams {
+			return nil, errInvalidStreamType.New()
+		}
+
+		var err error
+		logger := log.FromContext(ctx)
+
+		for attempt := uint(0); attempt <= callOpts.max; attempt++ {
+			retryTimeout := callOpts.timeout
+			err = waitRetryBackoff(ctx, retryTimeout, attempt)
+			if err != nil {
+				logger.WithError(err).Debug("An unexpected error occurred while in the timeout for the next request retry")
+				return nil, err
+			}
+
+			var clientStream grpc.ClientStream
+			callCtx := context.WithValue(ctx, attemptMetadataKey, attempt)
+			clientStream, err = streamer(callCtx, desc, cc, method, opts...)
+			if err == nil {
+				retriableStream := &retriableStreamClient{
+					ClientStream: clientStream,
+					callOpts:     callOpts,
+					parentCtx:    ctx,
+					streamerCall: func(ctx context.Context) (grpc.ClientStream, error) {
+						return streamer(ctx, desc, cc, method, opts...)
+					},
+				}
+				return retriableStream, nil
+			}
+
+			if !isRetriable(err, callOpts) {
+				return clientStream, err
+			}
+		}
+		return nil, err
+	}
+}
+
+func evaluateClientOpt(opts ...Option) *options {
+	optCopy := &options{}
+	*optCopy = *defaultOptions
+
+	for _, f := range opts {
+		f(optCopy)
+	}
+	return optCopy
+}
+
+func isRetriable(err error, opt *options) bool {
+	if err == nil {
+		return false
+	}
+	for _, check := range opt.validators {
+		if check(err) {
+			return true
+		}
+	}
+	return false
+}
+
+func waitRetryBackoff(ctx context.Context, timeout time.Duration, attempt uint) error {
+	if attempt == 0 {
+		return nil
+	}
+
+	log.FromContext(ctx).WithField("attempt", attempt).WithField("timeout", timeout).Info("Failed request, waiting until next attempt")
+	timer := time.NewTicker(timeout)
+	select {
+	case <-ctx.Done():
+		timer.Stop()
+		return ctx.Err()
+	case <-timer.C:
+	}
+	return nil
+}
+
+func hasRateLimitHeaders(md metadata.MD) bool {
+	return md != nil &&
+		len(md.Get(headerXRateLimit)) > 0 &&
+		len(md.Get(headerXRateAvailable)) > 0 &&
+		len(md.Get(headerXRateReset)) > 0 &&
+		len(md.Get(headerXRateRetry)) > 0
+}
+
+func getHeaderLimiterTimeout(ctx context.Context, md metadata.MD, opt *options) time.Duration {
+	if !hasRateLimitHeaders(md) {
+		return -1
+	}
+	reset, _ := strconv.Atoi(md.Get(headerXRateReset)[0])
+	timeout := time.Duration(reset) * time.Second
+	if opt.jitter > 0 {
+		rand.Seed(time.Now().UnixNano())
```

Comment:
I don't think it's our job to be concerned about implementation details of `backoffutils`. I suppose it makes sense to seed the PRNG once when the program starts, but that's also not our concern here.

Reply:
The `backoffutils` implementation is just a two liner but since we already import it I didn't see a reason to not use it. The `rand.Seed` was originally here since its okay to use it in a concurrent setting but I agree that it might be best to leave it where the program starts so I moved it to the `commands/root.go` file.